### DISCUSSION TITLE
How to provide multiple values with the same type in injection dependency?

### DISCUSSION THREAD

**Discussion by danuvian** [activej/activej#159] [Q&A]
I am testing how to use your dependency injection API, and I wanted to try to have a module define several String properties that can be injected into another class.

public class ConfigModule extends AbstractModule {

    @Provides
    String dbName() {
      return "exampleDB";
    }

    @Provides
    String dbUrl() {
        return "jdbc:mydb://localhost:1234" + dbName();
    }

    @Provides
    String dbUsername() {
        return "username";
    }

    @Provides
    String dbPassword() {
        return "pass";
    }
}

public class ServiceModule extends AbstractModule {
    @Provides
    Service service() {
        return new Service();
    }
}

public class Service {
    @Inject
    String dbName;

    @Inject
    String dbUrl;

    @Inject 
    String dbUsername;

    @Inject 
    String dbPassword;
}

If I call try to do the dependency injection like this:
public static void main(String[] args) {
        var injector = Injector.of(new ConfigModule(), new ServiceModule());
        var service = injector.getInstance(Service.class);
}

I get an exception:
Exception in thread "main" io.activej.inject.binding.DIException: Duplicate bindings for key String

Is there a way to do this with ActiveJ? If so, can you list out some sample examples here?

Thank you.

**Comment by eduard-vasinskyi** [comment]
Hi, @danuvian

Yes, you can provide objects of the same type but you need to use a qualifier to differentiate between them.

ActiveJ Inject has a notion of a [Key](https://github.com/activej/activej/blob/master/core-inject/src/main/java/io/activej/inject/Key.java) which is a `Type` + some qualifier to differentiate keys of the same type. You can read more about keys [here](https://activej.io/inject/basics#key). A qualifier can be an arbitrary `Object`, although for your needs you can use a built-in [@Named](https://github.com/activej/activej/blob/master/core-inject/src/main/java/io/activej/inject/annotation/Named.java) annotation which takes up a `String` qualifier (string should be unique within the same type).

You can look at the [ReflectionExample](https://github.com/activej/activej/blob/master/examples/core/inject/src/main/java/ReflectionExample.java) which uses different qualifiers as well as a custom qualifier annotation. 

Here is how your `ConfigModule` may look like after adding `@Named` annotations:
```java
public static class ConfigModule extends AbstractModule {
 
  @Provides
  @Named("DB Name")
  String dbName() {
     return "exampleDB";
  }
 
  @Provides
  @Named("DB URL")
  String dbUrl() {
     return "jdbc:mydb://localhost:1234" + dbName();
  }
 
  @Provides
  @Named("DB Username")
  String dbUsername() {
     return "username";
  }
 
  @Provides
  @Named("DB Password")
  String dbPassword() {
     return "pass";
  }
}
```

You would also need to modify a `Service` class to use `@Named` annotations as well:
```java
@Inject
public static class Service {
  @Inject
  @Named("DB Name")
  String dbName;
 
  @Inject
  @Named("DB URL")
  String dbUrl;
 
  @Inject
  @Named("DB Username")
  String dbUsername;
 
  @Inject
  @Named("DB Password")
  String dbPassword;
}
```

Notice that I’ve added an `@Inject` annotation on top of the Service class. This will help later when instances are being injected.

To inject instances to the Service class we need to rewrite a `ServiceModule` class so that it does not create an instance of the Service class explicitly, but rather notifies the Injector about the Service class. This way the `Injector` will create an instance of the Service class when asked and will also inject all of the required instances (those `String` properties for the database). You can read more about instance injecting [here](https://activej.io/inject/injecting-instances).

Here is how ServiceModule may look like (we simply bind the Service class): 
```java
public static class ServiceModule extends AbstractModule {
  @Override
  protected void configure() {
     bind(Service.class);
  }
}
```

 
 
Please let me know if you have any more questions regarding ActiveJ :)

_PS. If you plan to provide String values like configuration properties, you can use our [ConfigModule](https://activej.io/boot/config) rather than creating your own._
